plugins/scaleflux/sfx-nvme.c

```diff
@@ -0,0 +1,877 @@
+#include <fcntl.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <linux/fs.h>
+#include <inttypes.h>
+#include <asm/byteorder.h>
+#include <sys/ioctl.h>
+#include <sys/sysinfo.h>
+
+#include "linux/nvme_ioctl.h"
+
+#include "nvme.h"
+#include "nvme-print.h"
+#include "nvme-ioctl.h"
+#include "nvme-status.h"
+#include "json.h"
+#include "plugin.h"
+
+#include "argconfig.h"
+#include "suffix.h"
+
+#define CREATE_CMD
+#include "sfx-nvme.h"
+
+#define SFX_PAGE_SHIFT                      12
+#define SECTOR_SHIFT						9
+
+#define SFX_GET_FREESPACE           _IOWR('N', 0x240, struct sfx_freespace_ctx)
+#define IDEMA_CAP(exp_GB) 			(((__u64)exp_GB - 50ULL) * 1953504ULL + 97696368ULL)
+
+
+enum {
+    SFX_LOG_LATENCY_READ_STATS  = 0xc1,
+    SFX_LOG_SMART               = 0xc2,
+    SFX_LOG_LATENCY_WRITE_STATS = 0xc3,
+    SFX_LOG_QUAL                = 0xc4,
+    SFX_LOG_MISMATCHLBA         = 0xc5,
+    SFX_LOG_MEDIA               = 0xc6,
+    SFX_LOG_BBT                 = 0xc7,
+    SFX_LOG_IDENTIFY            = 0xcc,
+    SFX_FEAT_ATOMIC             = 0x01,
+};
+
+enum sfx_nvme_admin_opcode {
+    nvme_admin_query_cap_info   = 0xd3,
+    nvme_admin_change_cap       = 0xd4,
+    nvme_admin_sfx_set_features = 0xd5,
+    nvme_admin_sfx_get_features = 0xd6,
+};
+
+struct sfx_freespace_ctx
+{
+    __u64 free_space;
+    __u64 phy_cap;     /* physical capacity, in unit of sector */
+    __u64 phy_space;   /* physical space considering OP, in unit of sector */
+    __u64 user_space;  /* user required space, in unit of sector*/
+    __u64 hw_used;     /* hw space used in 4K */
+    __u64 app_written; /* app data written in 4K */
+};
+
+struct nvme_capacity_info {
+    __u64 lba_sec_sz;
+    __u64 phy_sec_sz;
+    __u64 used_space;
+    __u64 free_space;
+};
+#pragma pack(push,1)
```

Comment:
I know I've taken similar before, but i'd rather not use Microsoft dialects anymore. The gcc style is "__attribute__((packed))".

Reply:
will do.